PubSub: Pass push config in create subscription method

sergioisidoro opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/google-cloud-ruby/issues) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea.
- [x] Follow the instructions in [CONTRIBUTING](CONTRIBUTING.md). Most importantly, ensure the tests and linter pass by running `bundle exec rake ci` in the gem subdirectory.
- [x] Update code documentation if necessary.

closes: #7599

sergioisidoro commented:
@googlebot I fixed it.

quartzmo commented on google-cloud-pubsub/lib/google/cloud/pubsub/service.rb:
```diff
@@ -175,11 +175,21 @@ def list_subscriptions options = {}
+                          elsif !options[:push_config].is_a? Google::Cloud::PubSub::V1::PushConfig
+                            raise ArgumentError "push_config is not a valid pubsub PushConfig object"
+                          else
+                            options[:push_config]
```
Would this be anything other than a `Hash`? Might be safer to explicitly test for that, and otherwise raise `ArgumentError` in the final `else` block.

sergioisidoro replied:
Although in the end this becomes a hash, I thought it would be better to pass the objects instead, and reject everything else (including the hash).

This if clause is a bit confusing, because the raise is in the condition above.
I'll restructure the if condition into a more orthodox one raising only on the `else` block.

sergioisidoro replied:
Thanks for pointing this out. There was an actual bug on the raise branch of this `if` that was discovered when the tests were added :)

quartzmo commented on google-cloud-pubsub/test/google/cloud/pubsub/topic_test.rb:
```diff
@@ -154,6 +154,34 @@
+    new_sub_name = "new-sub-#{Time.now.to_i}"
+    endpoint = "http://foo.bar/baz"
+
+    push_config = Google::Cloud::PubSub::Subscription::PushConfig.new
```
Just noting that this currently `@private` constructor would need to be publicly documented.

sergioisidoro commented:
@quartzmo Sorry, I'm unable to run the acceptance tests. 

It seems the tests need a wide range of permissions, and I would need to setup a new Gcloud project with an account with massive permission rights just to test this.
eg. Why do pubsub tests require `storage.buckets.create` ?
```
Google::Apis::ClientError: forbidden: sergio-acceptance-tests@xxxxx.iam.gserviceaccount.com does not have storage.buckets.create access to the Google Cloud project.
```

sergioisidoro commented:
- Added documentations and examples
- Added an acceptance test (and ran it, everything seemed to went fine)
- Changed the if/else structure to be clear and added tests - (https://github.com/googleapis/google-cloud-ruby/pull/7600/files/1d9dbd206af531334df320feda8a03526147e3a4#diff-d0b5c7f9eb58f0018f1b656372c7a119)

quartzmo commented on google-cloud-pubsub/lib/google/cloud/pubsub/subscription/push_config.rb:
```diff
@@ -43,9 +43,24 @@ class Subscription
+          # Creates an empty configuration for a push subscription.
+          #
+          # @return [PushConfig]
           def initialize
```
Feel free to add params accepting the class attributes to `initialize` (keyword args for optional attributes) if you want, for convenience.

sergioisidoro commented:
* Added param documentations
* Added constructor arguments to push config, a basic test, and some documentation.
* Removed the mentioned non exposed API object.

Let me know what you think @quartzmo 

PS: Some linter rules are quite strict. I spent way too much time shuffling if/else clauses to match the linter suggestions. Maybe relaxing some of these would be nice.

sergioisidoro commented on google-cloud-pubsub/lib/google/cloud/pubsub/subscription/push_config.rb:
```diff
@@ -43,11 +43,33 @@ class Subscription
             @grpc = Google::Cloud::PubSub::V1::PushConfig.new
+
+            self.endpoint = endpoint unless endpoint.nil?
+            set_oidc_token auth_email, auth_audience if auth_email && auth_audience
```
Maybe we should throw an error if one, but not the other argument is provided. 
This silent failure might cause some head scratching.

quartzmo reviewed (commented):
@sergioisidoro This is looking good to me. I added just two very minor comments. Thanks for your patience and hard work on this change, I think it is a much needed addition.

sergioisidoro commented on google-cloud-pubsub/lib/google/cloud/pubsub/service.rb:
```diff
@@ -175,11 +175,18 @@ def list_subscriptions options = {}
+                          unless options[:push_config].is_a? Google::Cloud::PubSub::Subscription::PushConfig
+                            raise ArgumentError, "push_config is not a valid pubsub PushConfig object"
+                          end
+                          options[:push_config].to_grpc
```
Good question. I assumed that the service is a semi-private interface, since it's not well documented, and uses non structured, non documented arguments (ie. keyword arguments in `options`). So in that case throwing an error here feels a bit unorthodox.

Also `options[:attributes]` is not passed forward from `topic.subscribe`.

IMHO, PushConfig construction should be handled upstream (eg. in Topic). Maybe in here `push_config` could even be taken out of the `options` and transformed into an argument on its own - it's quite strange that we're providing constructor parameters of `PushConfig` in the service `options`.

But that's a breaking change. 

PR is already getting quite large. I wonder if we should try to merge something here, and try to address this somewhere else. My suggestion would be that in another PR we would:

- Add a deprecation warning to the options[:endpoint] and options[:attributes] in favour of `PushConfig` object construction in the `Service` class.
- Move the construction of the `PushConfig` upstream to `Topic`, and use `to_grpc` in the `Service` class
- Maybe add `push_auth_email`, `push_auth_audience`, and `push_attributes` to the arguments of `topic.subscribe`

At least we could add an Issue to track this situation.

quartzmo replied:
> Also options[:attributes] is not passed forward from topic.subscribe.

Thanks for pointing this out. In that case, it's not a problem.

> I assumed that the service is a semi-private interface, since it's not well documented, and uses non structured, non documented arguments (ie. keyword arguments in options)

Yes, `Service` is private. I think in a separate PR I will convert `options` to separate keyword args, for more clarity.

> This PR is already getting quite large. I wonder if we should try to merge something here, and try to address this somewhere else. 

Agreed! This is great, let's merge it!

quartzmo commented:
Thanks again @sergioisidoro!

quartzmo commented:
I'll post here once this change has been released, most likely Monday or Tuesday.
